test/fake_http_server.cc

```diff
@@ -87,38 +98,86 @@ void FakeServer::TerminateAllStreams() {
+    auto stream_it = path_streams.find(request.destination);
+    if (stream_it != path_streams.end()) {
+      auto& stream = stream_it->second;
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
         {"Content-Type", "text/plain"},
-    }));
-
-    // Create a queue for this watcher and add to the stream.
-    std::queue<std::string> my_queue;
-    stream.AddQueue(&my_queue);
-
-    // For every new string on my queue, send to the client.  The
-    // empty string indicates that we should terminate the stream.
-    while (true) {
-      std::string s = stream.GetNextResponse(&my_queue);
-      if (s.empty()) {
-        break;
+      }));
+
+      // Create a queue for this watcher and add to the stream.
+      std::queue<std::string> my_queue;
+      stream.AddQueue(&my_queue);
+
+      // For every new string on my queue, send to the client.  The
+      // empty string indicates that we should terminate the stream.
+      while (true) {
+        std::string s = stream.GetNextResponse(&my_queue);
+        if (s.empty()) {
+          break;
+        }
+        connection->write(s);
       }
-      connection->write(s);
+      return;
+    }
+
+    auto it = get_handlers.find(request.destination);
+    if (it != get_handlers.end()) {
+      const auto& handler = it->second;
+      std::map<std::string, std::string> headers;
+      for (const auto& h : request.headers) {
+        headers[h.name] = h.value;
+      }
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
+        {"Content-Type", "text/plain"},
+      }));
+      connection->write(handler(request.destination, headers));
+      return;
     }
-    return;
   }
 
-  auto it = path_responses.find(request.destination);
-  if (it != path_responses.end()) {
-    connection->set_status(Server::connection::ok);
-    connection->set_headers(std::map<std::string, std::string>({
+  if (request.method == "POST") {
+    auto it = post_handlers.find(request.destination);
+    if (it != post_handlers.end()) {
+      const auto& handler = it->second;
+      std::map<std::string, std::string> headers;
+      for (const auto& h : request.headers) {
+        headers[h.name] = h.value;
+      }
+      std::mutex body_mutex;
+      std::string body;
+      std::condition_variable body_retrieved;
+      std::unique_lock<std::mutex> body_lock(body_mutex);
+      connection->read([&](Server::connection::input_range range,
+                           boost::system::error_code error,
+                           size_t size,
+                           Server::connection_ptr conn) {
+        {
+          std::lock_guard<std::mutex> lock(body_mutex);
+          body.append(range.begin(), range.end());
+        }
+        body_retrieved.notify_all();
+      });
+      bool body_status = body_retrieved.wait_for(
+          body_lock, std::chrono::seconds(5),
```

Comment:
Looking back, this should have probably been a named constant... Not a big deal, just noting it for future work.